File: ml-agents-envs~/mlagents_dots_envs/test/test_shared_memory.py
import os
import tempfile
import numpy as np
from mlagents_dots_envs.shared_memory.base_shared_memory import BaseSharedMemory


def test_shared_memory():
    file_path = os.path.join(tempfile.gettempdir(), "ml-agents", "test")

    sm0 = BaseSharedMemory("test", True, 256)
    sm1 = BaseSharedMemory("test", False)

    new_offset = sm0.set_int(0, 3)
    val, off = sm1.get_int(0)
    assert off == new_offset
    assert val == 3

    new_offset = sm0.set_float(0, 3)
    val, off = sm1.get_float(0)
    assert off == new_offset
    assert val == 3

    new_offset = sm0.set_bool(0, True)
    val, off = sm1.get_bool(0)
    assert off == new_offset
    assert val

    sm0.set_string(0, "foo")
    assert sm1.get_string(0)[0] == "foo"
    off = sm1.set_string(0, "bar")
    assert sm0.get_string(0)[0] == "bar"
    assert off == sm0.get_string(0)[1]

    src = np.array([1, 2, 3, 4], dtype=np.float32)
    sm0.set_ndarray(0, src)
    dst = sm1.get_ndarray(0, (4,), np.float32)
    assert np.array_equal(src, dst)

    sm0.close()
    assert os.path.exists(file_path)
    sm1.delete()
    assert not os.path.exists(file_path)

File: Tests/Editor/TestSharedMemory.cs
using NUnit.Framework;
using UnityEngine;
using System.IO;
using Unity.AI.MLAgents;
using Unity.Collections;
using Unity.Mathematics;

namespace Unity.AI.MLAgents.Tests.Editor
{
    public class TestSharedMemory
    {
        [Test]
        public void TestBaseSharedMemory()
        {
            var filePath = Path.Combine(Path.GetTempPath(), "ml-agents", "test");
            File.Delete(filePath);

            var sm0 = new BaseSharedMemory("test", true, 256);
            var sm1 = new BaseSharedMemory("test", false);

            var newOffset = sm0.SetInt(0, 3);
            Assert.AreEqual(newOffset, sizeof(int));
            Assert.AreEqual(3, sm1.GetInt(0));

            newOffset = sm0.SetFloat(0, 3f);
            Assert.AreEqual(newOffset, sizeof(float));
            Assert.AreEqual(3f, sm1.GetFloat(0));

            newOffset = sm0.SetBool(0, true);
            Assert.AreEqual(newOffset, sizeof(bool));
            Assert.AreEqual(true, sm1.GetBool(0));

            newOffset = sm0.SetString(0, "foo");
            Assert.AreEqual("foo", sm1.GetString(0));
            sm1.SetString(newOffset, "bar");
            Assert.AreEqual("bar", sm0.GetString(newOffset));

            Assert.AreEqual(sm0.GetBytes(0, 8), sm1.GetBytes(0, 8));

            var src = new NativeArray<int4>(3, Allocator.Temp);
            src[1] = new int4(1, 2, 3, 4);
            sm0.SetArray(0, src, 3 * 16);
            var dst = new NativeArray<int4>(3, Allocator.Temp);
            sm1.GetArray(0, dst, 3 * 16);
            Assert.AreEqual(dst[1], new int4(1, 2, 3, 4));
            src.Dispose();
            dst.Dispose();

            sm0.Close();

            Assert.True(File.Exists(filePath));
            sm1.Delete();
            Assert.False(File.Exists(filePath));
        }
    }
}
